### DISCUSSION TITLE
timeout not respected with socks5 proxy

### DISCUSSION THREAD

**Discussion by a14n** [encode/httpx#2951] [Potential Issue]
With `httpx[socks]==0.23.3`and the following code an `httpx.ReadTimeout` is raised only with https proxy and not with socks5 proxy. Is it an known issue?

 ```py
async def main():
    async_client = httpx.AsyncClient(
#        proxies="https://my-https-proxy:443",
        proxies="socks5://my-socks5-proxy:3128",
        follow_redirects=True,
        timeout=1,
    )
    start = time.time_ns()
    try:
        response = await async_client.get(
            "http://www.protesa.net/",
            follow_redirects=False,
        )
        print("status", response.status_code)
    except Exception as e:
        print("ex", type(e))
  #      traceback.print_exc()
    finally:
        print(f"{time.time_ns() - start}ns")
    await async_client.aclose()

if __name__ == '__main__':
    asyncio.run(main())
```

The output for socks5 proxy is:

```
ex <class 'socksio.exceptions.ProtocolError'>
59288033291ns
```

The output for http proxy is:

```
ex <class 'httpx.ReadTimeout'>
1345715709ns
```

**Comment by dmittov** [comment]
I have the same issue.
I see `httpx` passes the timeout to `httpcore`. And then it's ignored.

So, it’s not a bug in `httpx` itself, but rather in one of its dependencies:
Issue: https://github.com/encode/httpcore/issues/1015
Fix: https://github.com/encode/httpcore/pull/1017

The fix works, but is not merged yet.

I understand working with a self-patched version is inconvenient, so I can suggest a workaround:

```diff
-response = await async_client.get("http://www.protesa.net/", follow_redirects=False)
+response = await asyncio.wait_for(
+    async_client.get("http://www.protesa.net/", follow_redirects=False),
+    timeout=1,
+)
```

It's not 100% the same, but can be a temporary solution in many cases.
